cache/cache.go

```diff
@@ -0,0 +1,328 @@
+/*
+Copyright 2024 The Flux authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cache
+
+import (
+	"fmt"
+	"sync"
+	"time"
+)
+
+// Cache is a thread-safe in-memory key/object store.
+type Cache struct {
+	*cache
+	// keyFunc is used to make the key for objects stored in and retrieved from items, and
+	// should be deterministic.
+	keyFunc KeyFunc
+}
+
+// Item is an item stored in the cache.
+type Item struct {
+	// Object is the item's object.
+	Object any
+	// Expiration is the item's expiration time.
+	Expiration int64
+}
+
+type cache struct {
+	// Items holds the elements in the cache.
+	Items map[string]Item
+	// MaxItems is the maximum number of items the cache can hold.
+	MaxItems int
+	mu       sync.RWMutex
+	janitor  *janitor
+	closed   bool
+}
+
+var _ Expirable = &Cache{}
+
+// New creates a new cache with the given configuration.
+func New(maxItems int, keyFunc KeyFunc, interval time.Duration) *Cache {
+	c := &cache{
+		Items:    make(map[string]Item),
+		MaxItems: maxItems,
+		janitor: &janitor{
+			interval: interval,
+			stop:     make(chan bool),
+		},
+	}
+
+	C := &Cache{cache: c, keyFunc: keyFunc}
+
+	if interval > 0 {
+		go c.janitor.run(c)
+	}
+
+	return C
+}
+
+func (c *Cache) Close() error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	if c.closed {
+		return fmt.Errorf("cache already closed")
+	}
+	c.janitor.stop <- true
+	c.closed = true
+	return nil
+}
+
+// Add an item to the cache, existing items will not be overwritten.
+// To overwrite existing items, use Update.
+// If the cache is full, Add will return an error.
+func (c *Cache) Add(object any) error {
+	key, err := c.keyFunc(object)
+	if err != nil {
+		return KeyError{object, err}
+	}
+
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	if c.closed {
+		return KeyError{object, fmt.Errorf("cache already closed")}
+	}
+	_, found := c.Items[key]
+	if found {
+		return KeyError{object, fmt.Errorf("key already exists")}
+	}
+
+	if c.MaxItems > 0 && len(c.Items) < c.MaxItems {
+		c.set(key, object)
+		return nil
+	}
+
+	return KeyError{object, fmt.Errorf("Cache is full")}
```

Comment:
Instead of erroring out, I think we should evict the key with the closest expiry.

Reply:
If I set an `expiration`, it should be guaranteed.